halo2-base/src/virtual_region/manager.rs

```diff
@@ -0,0 +1,16 @@
+use crate::ff::Field;
+use crate::halo2_proofs::circuit::Region;
+
+/// A virtual region manager is responsible for managing a virtual region and assigning the
+/// virtual region to a physical Halo2 region.
+///
+pub trait VirtualRegionManager<F: Field> {
+    /// The Halo2 config with associated columns and gates describing the physical Halo2 region
+    /// that this virtual region manager is responsible for.
+    type Config: Clone;
```

Comment:
That would not support the case of multiple different gates on the same columns.

Reply:
`VirtualRegionManager` owns `Config` but it doesn't have to own `Columns`. We can just keep references of `Columns` in `Config`. The point is to determine the columns of the virtual region during `Circuit::configure`. `fn configure` in the trait is confusing, `fn initialize` might be better.

Reply:
Can you explain again what the intended flow is here? Is `initialize` called during `Circuit::configure` or Circuit::synthesize`? I think it's impossible to call during `Circuit::configure` because there all that is passed in is `configure_with_params(meta, params)`. 

So then the only time you would access the columns would be in `synthesize`, but then it's effectively the same as providing `Self::Config` as an argument to `assign_raw`?

Reply:
Ah you're right. I forgot that `Circuit` doesn't own `Circuit::Config`(now I don't understand why..). So you can only access the columns in `synthesize`.

The point here is, the current interface looks like you could call `assign_raw` multiple times with different `Config` but actually it's unexpected. We add `assigned: Arc<OnceLock()>` to prevent it but you only know it when reading internal implementation. If you bind `Config` with `VirtualRegionManager`, semantically `assign_raw` should be only called once or at least it's idempotent.

Reply:
yea I understand, but not sure how to do it easily without making devex annoying because of this `Circuit::configure` thing.

Reply:
I think it should almost identical?
Before:
```
vrm.assign_raw(config, region);
```
After:
```
vrm.initialize(config);
vrm.assign_raw(region);
```

Reply:
oh I see, you just want to make sure `initialize` is only done once. sure we can do that